### PR TITLE
ADF-654: Fix persisting ACLs when applying a huge number of permissions at once

### PR DESCRIPTION
**Pull request by hectoras**
Persist  new permissions using multiple `INSERT`s to prevent exceeding the maximum MySQL parameter count per query.

These changes use a batch size of **20K** permissions per `INSERT`. Given each permission stores values for 3 columns per row (i.e. per resource), we get `20K * 3 = 60K` parameters per query, which is less than the observed MySQL limit of **65K**. The "chunk size" is stored as a constant, but it probably can be moved to a config parameter if preferred.

Note the limit for MariaDB (used locally for development) seems higher than that (some sources cite ~1000 parameters, but I was able to go up to 250K parameters locally before giving up trying to reproduce the error). By the way, this change  makes TAO to issue various "smaller" queries (of 20K rows each) instead of a single, big one, which should be more resource friendly on the PHP side.

I'm also adding a couple of logger calls to improve observability, but that can be removed if you think that is not necessary.

**Associated Jira issue:** [ADF-654](https://oat-sa.atlassian.net/browse/ADF-654)

---
TODO
- [x] Review & fix things pointed out during code review
- [x] Check existing unit tests and adapt and/or add new ones
- [x] Wrap database operations in a DBAL transaction

### CODE REVIEW

File: model/DataBaseAccess.php [hunk 1]
```diff
@@ -177,7 +178,28 @@ public function addMultiplePermissions(array $permissionData)
                 }
             }
         }
-        $this->getPersistence()->insertMultiple(self::TABLE_PRIVILEGES_NAME, $insert);
+
+        $this->getLogger()->debug(
+            "Processing {count} permission inserts in {chunks} chunks",
```
Comment (shpran) [inline model/DataBaseAccess.php:183]: Please, use single quotes.
When you are using double quotes - you are checking the string for possible variables.

File: model/DataBaseAccess.php [hunk 2]
```diff
@@ -177,7 +178,28 @@ public function addMultiplePermissions(array $permissionData)
...
+            [
+                'count' => count($insert),
+                'chunks' => ceil(count($insert)/self::INSERT_CHUNK_SIZE)
```
Comment (shpran) [inline model/DataBaseAccess.php:186]: Store `count($insert)` in a separate variable and use it instead.

File: model/DataBaseAccess.php [hunk 3]
```diff
@@ -177,7 +178,28 @@ public function addMultiplePermissions(array $permissionData)
...
+            ]
+        );
+
+        foreach(array_chunk($insert, self::INSERT_CHUNK_SIZE) as $index => $batch) {
+            $this->getLogger()->debug(
+                "Processing chunk {index}/{total} with {items} ACL entries",
```
Comment (shpran) [inline model/DataBaseAccess.php:192]: Please, use single quotes.
When you are using double quotes - you are checking the string for possible variables.

File: model/DataBaseAccess.php [hunk 4]
```diff
@@ -177,7 +178,28 @@ public function addMultiplePermissions(array $permissionData)
...
+                [
+                    'index' => $index + 1,
+                    'total' => ceil(count($insert)/self::INSERT_CHUNK_SIZE),
```
Comment (shpran) [inline model/DataBaseAccess.php:195]: Use separate variable for `count($insert)` instead of counting on every iteration.

File: model/DataBaseAccess.php [hunk 5]
```diff
@@ -177,7 +178,28 @@ public function addMultiplePermissions(array $permissionData)
...
+        $this->getLogger()->debug(
+            "Processing {count} permission inserts in {chunks} chunks",
+            [
+                'count' => count($insert),
+                'chunks' => ceil(count($insert)/self::INSERT_CHUNK_SIZE)
+            ]
+        );
+
+        foreach(array_chunk($insert, self::INSERT_CHUNK_SIZE) as $index => $batch) {
+            $this->getLogger()->debug(
```
Comment (shpran) [inline model/DataBaseAccess.php:191]: It's not so critical, but maybe we can store `getLogger` in a separate variable or cache this call?

```php
    public function getLogger()
    {
        if ($this->logger instanceof LoggerInterface) {
            return $this->logger;
        }
        if ($this instanceof ServiceLocatorAwareInterface) {
            $logger = $this->getServiceLocator()->get(LoggerService::SERVICE_ID);
        } else {
            $logger = ServiceManager::getServiceManager()->get(LoggerService::SERVICE_ID);
        }
        return ($logger instanceof LoggerInterface) ? $logger : new NullLogger();
    }
```

As you can see, if logger is an instance of `LoggerInterface` - we are good. But in other cases, we will use `ServiceLocator`/`ServiceManager` to get the necessary service. And this might impact a bit on performance.

But as I said, this is not so critical, and you can skip this one 🙂😅

File: model/DataBaseAccess.php [hunk 6]
```diff
@@ -177,7 +178,28 @@ public function addMultiplePermissions(array $permissionData)
...
+
+        $this->getLogger()->debug(
+            "Processing {count} permission inserts in {chunks} chunks",
+            [
+                'count' => count($insert),
+                'chunks' => ceil(count($insert)/self::INSERT_CHUNK_SIZE)
+            ]
+        );
+
+        foreach(array_chunk($insert, self::INSERT_CHUNK_SIZE) as $index => $batch) {
```
Comment (shpran) [inline model/DataBaseAccess.php:190]:
```suggestion
        foreach (array_chunk($insert, self::INSERT_CHUNK_SIZE) as $index => $batch) {
```

File: model/DataBaseAccess.php (hunk 1 above)
Comment (hectoras) [reply model/DataBaseAccess.php]: Very true :facepalm: Changed in 39763796a360382af747a8883cc311b13dfe8ddc, thanks for spotting this :+1:

File: model/DataBaseAccess.php (hunk 3 above)
Comment (hectoras) [reply model/DataBaseAccess.php]: Very true :facepalm: Changed in 39763796a360382af747a8883cc311b13dfe8ddc, thanks for spotting this :+1:

File: model/DataBaseAccess.php [hunk 7]
```diff
@@ -48,6 +48,7 @@ class DataBaseAccess extends ConfigurableService
     const COLUMN_PRIVILEGE = 'privilege';
     const TABLE_PRIVILEGES_NAME = 'data_privileges';
     const INDEX_RESOURCE_ID = 'data_privileges_resource_id_index';
+    const INSERT_CHUNK_SIZE = 20000;
```
[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: Do you think we should set other constants in this class explicitly as `public` as well or leave them as they are now?

cc/ @gabrielfs7

File: model/DataBaseAccess.php (hunk 2 above)
Comment (hectoras) [reply model/DataBaseAccess.php]: Changed in 0c6e230eaa4b7ffc1a0225ae00c07c1dd8942614 along with some linter errors. Thanks!

File: model/DataBaseAccess.php (hunk 4 above)
Comment (hectoras) [reply model/DataBaseAccess.php]: Missed that, changed in 0c6e230eaa4b7ffc1a0225ae00c07c1dd8942614 Thanks!

File: model/DataBaseAccess.php (hunk 6 above)
Comment (hectoras) [reply model/DataBaseAccess.php]: Already fixed in 0c6e230eaa4b7ffc1a0225ae00c07c1dd8942614 (I should have run the linter before pushing the branch :sweat_smile: )

File: model/DataBaseAccess.php (hunk 5 above)
Comment (hectoras) [reply model/DataBaseAccess.php]: That's a good point, changed in b95ca785c1c655a49c2e1fa8725e365b78b54a99 to use a single reference to the logger through the method. Thanks :slightly_smiling_face:

File: model/DataBaseAccess.php [hunk 8]
```diff
@@ -177,7 +178,28 @@ public function addMultiplePermissions(array $permissionData)
                 }
             }
         }
-        $this->getPersistence()->insertMultiple(self::TABLE_PRIVILEGES_NAME, $insert);
+
+        $this->getLogger()->debug(
```
[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: Moved new code into a separate method, and wrapped all insertions with a single transactional() call in 60a7e8314c5ae87087cd8b9e659beeae2886ad7e. I totally missed about the possibility of wrapping the insertions into a transaction, thanks!

I'll update unit tests next.

[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: Tests updated in 9694d05fb318aa5398fe392ee9a5b1150d74e5e2 (then, 44efff00c386b731c528f51f1e101cd1c5cbb9fa removes the additional echo I accidentally committed).

As these tests are checking edge cases with sets of more than 20K permissions, running them is slow and takes a lot of memory. @gabrielfs7 @shpran wdyt about setting `INSERT_CHUNK_SIZE` as a parameter of `DataBaseAccess` instead of a constant? We may allow passing it as an option to the constructor (inherited from the `Configurable` class), so it is set to 20K by default but initialized to some lower value for tests.

[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: We've agreed to replace the `INSERT_CHUNK_SIZE` class constant with a variable so we can lower the number of rows per query for unit tests using a  setter method. The new variable will have a default value of  20K rows.

[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: Changed here :point_right: b81293facb914a03a0e3f2e401e6abc58897da0f

File: model/DataBaseAccess.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: After last changes this constant is now a private variable.

As for the existing ones, `COLUMN_USER_ID` and `COLUMN_PRIVILEGE` are used only from unit tests, but the other 3 are also used from the updater script. Therefore, I'd opt for setting all of them explicitly to public to keep the current semantics. wdyt?

File: model/DataBaseAccess.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: New code has been moved to a private method, queries now run inside a transaction, and we've already updated unit tests and simplified them (i.e. eb507d0e7496970760cd09c3deb10c3dbce8e17e). Therefore, closing this thread as I think we've already handled all required points for it.

**Review by gabrielfs7 — APPROVED** [review]
Good job :rocket

- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: model/DataBaseAccess.php [hunk 9]
```diff
@@ -385,4 +392,41 @@ public function removeTables()
...
+            'Processing {count} permission inserts in {chunks} chunks',
+            [
+                'count' => count($insert),
+                'chunks' => ceil($insertCount / self::INSERT_CHUNK_SIZE)
+            ]
+        );
+
+        if(empty($insert)) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: Yes, I missed that one yesterday. I've committed your suggested change in 502b7f780eba1f009a2a118f15883e9a6e3d63e0. Thanks!

File: test/unit/model/DataBaseAccessTest.php [hunk 10]
```diff
@@ -172,23 +219,111 @@ public function testGetPermissions($userIds, array $resourceIds)
...
+    public function testAddMultiplePermissions($numEvents, array $permissionData)
+    {
+        $numInserts = (int) ceil($numEvents / self::INSERT_CHUNK_SIZE);
+
+        $persistenceMock = $this->getPersistenceMockForSetOps($numInserts);
+        $this->setPersistence($this->instance, $persistenceMock);
+        $this->setEventManager($this->instance, $numEvents);
+
+        $this->instance->addMultiplePermissions($permissionData);
+    }
```
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: I think both things have been already fixed in eb507d0e7496970760cd09c3deb10c3dbce8e17e

File: test/unit/model/DataBaseAccessTest.php [hunk 11]
```diff
@@ -92,6 +100,46 @@ public function getPersistenceMock($queryParams, $queryFixture, $resultFixture)
         return $persistenceMock;
     }
 
+    /**
+     * Return a persistence Mock object supporting transactional() & insertMultiple()
+     * @return MockObject
+     */
+    public function getPersistenceMockForSetOps(int $numInsertMultipleCalls) {
```
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: Yesterday we've already removed this method altogether, and moved initialization to `setUp()` instead.

As for PSR-12, right now phpcs reports issues with constant visibility and lines exceeding 120 chars (I'll fix both as part of this pull request).

File: test/unit/model/DataBaseAccessTest.php [hunk 12]
```diff
@@ -172,23 +219,111 @@ public function testGetPermissions($userIds, array $resourceIds)
...
+            ->expects($this->exactly($numTriggeredEvents))
+            ->method('trigger')
+            ->with($this->isInstanceOf(DacAddedEvent::class));
+
+        $locator = $this->getServiceLocatorMock([
+            EventManager::SERVICE_ID => $eventManager
+        ]);
+
+        $instance->setServiceManager($locator);
+    }
```
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: Yesterday we've already moved this to `setUp()` ([here](https://github.com/oat-sa/extension-tao-dac-simple/pull/183/files#diff-2f691acdadccce78c1de5e057aa90ad8713db94e6f302259f929c6dd182204c2R56)), I think this thread can be closed now, right?

File: test/unit/model/DataBaseAccessTest.php [hunk 13]
```diff
@@ -172,23 +219,111 @@ public function testGetPermissions($userIds, array $resourceIds)
...
+    private function setPersistence(DataBaseAccess $instance, $persistenceMock)
     {
-        $reflectionClass = new \ReflectionClass(get_class($instance));
-        $persistence = $reflectionClass->getProperty('persistence');
-        $persistence->setAccessible(true);
-        $persistence->setValue($instance, $persistenceMock);
+        $reflector = new \ReflectionProperty(get_class($instance), 'persistence');
+        $reflector->setAccessible(true);
+        $reflector->setValue($instance, $persistenceMock);
+    }
```
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: This method was part of the former unit test. At first I just changed `ReflectionClass` by `ReflectionProperty` and I think I initially moved it to the end of the test file as well.

However, we've already moved this to `setUp()` in eb507d0e7496970760cd09c3deb10c3dbce8e17e, so I think this thread can be closed, right?

File: test/unit/model/DataBaseAccessTest.php [hunk 14]
```diff
@@ -172,23 +237,150 @@ public function testGetPermissions($userIds, array $resourceIds)
...
+            'permissionData' => [
+                [
+                    'resource' => $this->getResourceMock((string)($resourceId)),
+                    'permissions' => $this->getPermissionsArrayGenerator($numPermissions)
+                ]
+            ],
+        ];
+    }
+
+    private function getPermissionsArrayGenerator(int $numPermissions): Generator
```
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: That's true, it makes the test harder to write and harder to follow. However, we've already changed the approach by allowing lowering the number of inserted rows per query instead of generating huge arrays of permissions.

This was used as a way to generate big sets of permissions using as less memory as possible for them (instead of building the full array before the test start) in order to test INSERTs were split correctly.

File: test/unit/model/DataBaseAccessTest.php [hunk 15]
```diff
@@ -40,34 +44,56 @@
...
+
     /**
      * @var DataBaseAccess
      */
-    protected $instance;
+    private $sut;
+
+    private $randomPermissionSets = [];
+
+    private static $knownPermissionTypes = ['GRANT', 'READ', 'WRITE'];
```
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: This did not change between test (it was used to improve readability a little bit), and could have been defined as a constant as well. However, we've already removed it when we removed tests for random permissions.

File: model/DataBaseAccess.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: As previously agreed on Slack, declaring these as `public const` in cc457f6173a9fd3bef46aceac53e096b2a93e480

File: test/unit/model/DataBaseAccessTest.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: Fixed PSR12 issues reported by `phpcs` in 29938e17d26ecd5e0cf907c81eae33b2c92fd553

**Comment by hectoras** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Pull request's target is not master
- [ ] Kitchen links are not exposed in the description
- [ ] Commits are following conventional commits

**Comment by AntonNeviarouski** [comment]
QA verified

File: model/DataBaseAccess.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply model/DataBaseAccess.php]: This was already changed, so I'm marking this thread as solved. Thanks!

File: test/unit/model/DataBaseAccessTest.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: This was already taken care of, so I'm marking this thread as solved. Thanks!

File: test/unit/model/DataBaseAccessTest.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: This was already taken care of, so I'm marking this thread as solved. Thanks!

File: test/unit/model/DataBaseAccessTest.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: This was already changed (`setPersisrtence()` is removed, and its logic moved to `setUp()`), so I'm marking this thread as solved. Thanks!

File: test/unit/model/DataBaseAccessTest.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: Closing this thread as well :)

File: test/unit/model/DataBaseAccessTest.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (hectoras) [reply test/unit/model/DataBaseAccessTest.php]: This was already changed, so I'm marking this thread as solved. Thanks!

**Comment by hectoras** [comment]
All things pointed out during the code review have been already taken care of, and QA has verified this is working as expected.

Therefore, merging this pull request.
